src/imp.rs

```diff
@@ -0,0 +1,288 @@
+///
+/// The design is close to
+/// https://github.com/apple/swift/tree/bc3189a2d265bf7728ea0cfeb55f032bfe5beaf1/lib/Syntax
+///
+/// `SyntaxNode` exists in two flavors:
+///   * owned (R = OwnedRoot<T>)
+///   * borrowed (R = RefRoot<'a, T>)
+///
+/// Borrowed `SyntaxNode` is `Copy`, but is parametrized over a lifetime,
+/// with a corresponding ergonomics hit.
+///
+/// Owned `SyntaxNode` is `Clone` (using `Arc::clone` under the hood) and
+/// is not parametrized over a lifetime. Note that because of the parent
+/// links `SyntaxNode` keeps all of its ancestors alive, and not only descendants,
+/// so keep an eye on memory leaks.
+///
+/// Methods like `parent` or `children` preserve the flavor (borrowed or owned)
+/// of nodes, but you can switch between them at any time using `.borrowed()`
+/// and `.owned()` methods. As a rule of thumb, when *processing* nodes, use
+/// borrowed version to avoid excessive Arc traffic, and, when *storing* nodes
+/// in data structures, use owned variant, to avoid dealing with lifetimes.
+///
+/// `SyntaxNode` have object identity equality and hash semantics.
+pub struct SyntaxNode<T: Types> {
+    // Created from `Arc`. The ref-count on root is equal to the number of live
+    // `TreePtr` which point into this tree.
+    root: *const SyntaxRoot<T>,
+    pub(crate) parent: Option<ParentData<T>>,
+    pub(crate) green: GreenNode<T>,
+    // replace with
+    // `children: [(TextUnit, AtomSetOnce<SyntaxNode<T><T>>)]`
+    // once we can have var-length structs.
+    // Perhaps even fold `TextUnit` into ptr using a spare bit?
+    pub(crate) children: Box<[SwapCell<TextUnit, SyntaxNode<T>>]>,
+}
+
+// Manualy send/sync impls due to `root: *const SyntaxRoot<T>` field.
+unsafe impl<T> Send for SyntaxNode<T>
+where
+    T: Types,
+    SyntaxNode<T>: Send,
+{
+}
+
+unsafe impl<T> Sync for SyntaxNode<T>
+where
+    T: Types,
+    SyntaxNode<T>: Send,
+{
+}
+
+/// Owned smart pointer for syntax Nodes.
+/// It can be used with any type implementing `TransparentNewType<SyntaxNode>`.
+// Conceptually, it also plays an `Arc<TreeRoot<T>>` role as well.
+pub struct TreePtr<T, N>
+where
+    T: Types,
+    N: TransparentNewType<Repr = SyntaxNode<T>>,
+{
+    pub(crate) inner: *const N,
+}
+
+/// A marker trait for transparent newtypes.
+///
+/// If you declare a struct like
+///
+/// ```no-run
+/// #[repr(transparent)]
+/// struct Wrapper(Inner)
+/// ```
+///
+/// it is safe to add
+///
+/// ```no-run
+/// unsafe impl TransparentNewType for Wrapper { type Repr = Inner; }
+/// ```
+///
+/// Implementing this trait allows one to cast safely between the wrapper and
+/// the underlying representation.
+pub unsafe trait TransparentNewType: Sized {
+    /// Underlyng representation of a newtype.
+    type Repr;
+    /// Cast the uderlying repr into a wrapper.
+    fn from_repr(repr: &Self::Repr) -> &Self {
+        assert!(mem::size_of::<Self>() == mem::size_of::<Self::Repr>());
+        unsafe { mem::transmute(repr) }
```

Comment:
Good catch!

Interestingly, while implementing this, I've origianlly written 

```
unsafe { &*(repr as * const Self::Repr as * const &Self) }
```

instead of

```
unsafe { &*(repr as * const Self::Repr as * const Self) }
```

and the first version is compiled. I've looked for a crate here, but hanven't found one.